tests/v2/validation/provisioning/k3s/etcd_node_delete_test.go

```diff
@@ -0,0 +1,162 @@
+package k3s
+
+import (
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	v1 "github.com/rancher/rancher/tests/framework/clients/rancher/v1"
+	"github.com/rancher/rancher/tests/framework/extensions/clusters"
+	"github.com/rancher/rancher/tests/framework/extensions/machinepools"
+	nodestat "github.com/rancher/rancher/tests/framework/extensions/nodes"
+	"github.com/rancher/rancher/tests/framework/extensions/workloads/pods"
+	"github.com/rancher/rancher/tests/framework/pkg/config"
+	"github.com/rancher/rancher/tests/framework/pkg/session"
+	"github.com/rancher/rancher/tests/v2/validation/provisioning"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	kwait "k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	defaultNS                = "fleet-default"
+	etcdLabel                = "rke.cattle.io/etcd-role"
+	clusterLabel             = "cluster.x-k8s.io/cluster-name"
+	machineSteveResourceType = "cluster.x-k8s.io.machine"
+)
+
+type EtcdNodeDeleteTestSuite struct {
+	suite.Suite
+	session               *session.Session
+	client                *rancher.Client
+	ns                    string
+	k3skubernetesVersions []string
+	cnis                  []string
+	providers             []string
+	nodesAndRoles         []machinepools.NodeRoles
+}
+
+func (e *EtcdNodeDeleteTestSuite) TearDownSuite() {
+	e.session.Cleanup()
+}
+
+func (e *EtcdNodeDeleteTestSuite) SetupSuite() {
+	testSession := session.NewSession()
+	e.session = testSession
+
+	e.ns = defaultNS
+
+	clustersConfig := new(provisioning.Config)
+	config.LoadConfig(provisioning.ConfigurationFileKey, clustersConfig)
+
+	e.k3skubernetesVersions = clustersConfig.K3SKubernetesVersions
+
+	e.cnis = clustersConfig.CNIs
+	e.providers = clustersConfig.Providers
+	e.nodesAndRoles = clustersConfig.NodesAndRoles
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(e.T(), err)
+
+	e.client = client
+}
+
+func (e *EtcdNodeDeleteTestSuite) isEtcdNodeRemoved(query url.Values, clusterName string, etcdNodeToDelete v1.SteveAPIObject) (int, error) {
```

Comment:
Can this be abstracted out to the extensions location? Since this is a test file, we usually have majority, if not all, of functions like this abstracted in the extensions folder.

Reply:
I would also change the function name, as it explicitly checks:
* is the specified etcd node deleted

but has logging for:
* has the etcd node been replaced

I believe that the replacement of the etcd node should be validated as well. Can you confirm, add this functionality, and update the function name to reflect this?

Reply:
agreed

Reply:
@markusewalker, @slickwarren, I've moved the function to extensions and made below changes to it.
I've changed the function name to `IsRKE1EtcdNodeReplaced` and `IsRKE2K3SEtcdNodeReplaced`. And to validate whether etcd node is deleted and replaced i'm doing two things:
- Once the etcd node is deleted this function polls to fetch nodes list and validates deleted etcd node is not present in the nodes list.
- Also this function now takes `numOfEtcdNodesBeforeDeletion` and once the etcd node is deleted and not present in the nodes list, it again counts etcd nodes and stores in `numOfEtcdNodesAfterDeletion` and return bool by equating the variables.